Possibility to use ObservableArray in Store instead of a dstore/Store delite

harbalk opened the pull request:
**This pull request is just made to have some feedback on the work I have done until now**. It is not complete for the moment.

The goal is to approach the current use of delite/Store with an array instead of a dstore/Store : https://github.com/ibm-js/delite/issues/278

What have been done : 
- Modified Store.js to be able to accept Array and ObservableArray
- Documentation to describe the use of delite/Store with an array in store property
- Unit tests on Store and StoreMap with arrays

What still remaining : 
- Possible renaming of variables and elements (https://github.com/ibm-js/delite/issues/308)
- Example of communication between the array and a server

wkeese commented:
I looked over this briefly.   It seems good.  The two main issues/questions I had are:

1. Looks like for observable arrays getItemByIdentity() is O(n) instead of O(1) and I wonder about the performance.

2. Having if()/else() branches in many functions is not very object oriented.   Perhaps instead you should have helper classes that wrap the observable array and/or wrap the store, to give a standard interface that delite/Store could call.   Not sure if that makes sense in this situation or not.

asudoh commented:
Thank you for working on this! In addition to the specific comments to the code I’m adding… I’m wondering, would we still need delite/Store interface if it’s not abstracting dstore? Can we just let widgets directly use ObservableArray directly if we [put it in decor](https://github.com/ibm-js/decor/pull/48)?

asudoh commented on Store.js:
```diff
@@ -140,14 +150,109 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+		 * Synchronously deliver change records of the Observe function for the Array
+		 * Do NOT deliver change records of the Observe function for Objects
+		 */
+		deliver: dcl.superCall(function (sup) {
```
Should we deliver the change records for array entries here, too?

harbalk replied:
For the first point it was related to keep Observable.observe returns or not we discussed.
For the second is this function expect to call the remove function of the observe return ?

asudoh replied:
Please refer to https://github.com/polymer/observe-js#observable. My comment here is about how `deliver()` (and `discardChagnes()` should work (rather than about `Observable.observe()` handle). `deliver()` should deliver all pending change records (also refer to http://arv.github.io/ecmascript-object-observe/#DeliverChangeRecords). `discardChanges()` should throw away all pending change records.

harbalk replied:
Ok I thought Observable.observe was returning a deliver method like ObservableArray.

harbalk replied:
Fixed in 6b8997245c31c217d70357b33780da019fc1d41a.

wkeese commented:
> would we still need delite/Store interface if it’s not abstracting dstore? Can we just let widgets directly use ObservableArray directly if we put it in decor?

I think the idea is that widgets can interface to either a dstore or an ObservableArray.

asudoh commented on Store.js:
```diff
@@ -140,14 +150,109 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+		observeCallback: function (changeRecords) {
+			for (var i = 0; i < changeRecords.length; i++) {
+				// array modified
+				if (Array.isArray(changeRecords[i].object) === true) {
```
Is there a possibility to have `this.store` as a two-dimensional array?

harbalk replied:
I did not think about that in my implementation. It may need some modifications.

asudoh commented on Store.js:
```diff
@@ -140,14 +150,109 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+		 * @param changeRecords - send by the Observe function
+		 */
+		observeCallbackItems: function (change) {
+			if (change.type === "add" || change.type === "update" || change.type === "delete") {
```
Change records of objects should be debounced, like [what ObservablePath does](https://github.com/ibm-js/liaison/blob/30f907ad879ff337b3cc42969d436b2543dc47a3/ObservablePath.js#L86-L96).

harbalk replied:
I'm not sure what you mean by that, can you give me more details please ?

asudoh replied:
Change record is created for every change in object. If multiple changes to an object happens in a single [event-loop](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop), multiple change records of the object is created and sent to `Observable.observe()` callback. In such scenario `this._itemUpdated()` needs to be run only once. Therefore calling `this._itemUpdated()` for every change record causes redundant calls.

harbalk replied:
Ok so I have to check if the item is not modified by another record before calling this._itemUpdated, Thank you.

asudoh replied:
I guess the `this.observeCallback()` (if called from [this line](https://github.com/ibm-js/delite/pull/410/files#diff-3306122b521ab15fecaabf6fdc73a3dbR232)) can group change records by target object, and call `this._itemUpdated()` just once for each object.

harbalk replied:
Fixed in 6b8997245c31c217d70357b33780da019fc1d41a. Thanks.

asudoh commented:
>> would we still need delite/Store interface if it’s not abstracting dstore? Can we just let widgets directly use ObservableArray directly if we put it in decor?

> I think the idea is that widgets can interface to either a dstore or an ObservableArray.

I see it in the code now, thank you for clarification.

asudoh commented on Store.js:
```diff
@@ -120,18 +127,21 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
-					if (this._attachedlistener) {
-						this.store.removeEventListener("customelement-attached", this._attachedlistener);
+				var collection;
+				if (Array.isArray(this.store) === false) {
```
As @wkeese mentioned in another place, would falsy check be enough here?

asudoh commented on Store.js:
```diff
@@ -140,14 +150,109 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+				} else {
+					for (var i = 0; i < this.store.length; i++) {
+						// affect the callback to the observe function if the item is observable
+						Observable.observe(this.store[i], this.observeCallback.bind(this));
```
What `Observable.observe()` returns should be kept somewhere so that it can be cleaned up as necessary (e.g. when the array item is removed from `this.store`, when the widget is being destroyed).

harbalk replied:
I thought about that, but I seems to me that the cost of keeping the result of the observe function for all the items was too high. I do not know what tou think about it ?

asudoh replied:
I see your concern. OTOH observations not being cleaned up property will cause memory leak when widgets are destroyed and/or `this.store` is modified even. So I'd be leaned toward keeping observation handle around anyway.

wkeese replied:
I agree.  You can release the old handles when `store` is changed or on detachedCallback.

harbalk replied:
Ok I will do the modification.

harbalk replied:
Fixed in 6b8997245c31c217d70357b33780da019fc1d41a. Thanks.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+		 * @param item
+		 * @private
+		 */
+		_isItemObservable: function (item) {
```
This function seems equivalent of `Observable.canObserve()`.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+		 * @param item
+		 * @private
+		 */
+		_isArrayObservable: function (item) {
```
This function seems equivalent of `ObservableArray.canObserve()`.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+					for (var i = 0; i < this.store.length; i++) {
+						if (this._isItemObservable(this.store[i])) {
+							this._hasItemObservable = true;
+							this._observeResults =  [];
```
Do you really want to reset `this._observeResults` every occurrence in the loop?

harbalk replied:
Of course not, I have had a test if(!this._observeResults) like under. Sorry I didn't push it.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+			if (!this._beingDiscarded) {
+				for (var i = 0; i < changeRecords.length; i++) {
+					// array modified
+					if (Array.isArray(changeRecords[i].object)) {
```
Please refer to: https://github.com/ibm-js/delite/pull/410#discussion_r31128748

harbalk replied:
I discussed about the possibility to use two dimensional arrays with @cjolif. We did not really find a use case, so I did not do the modifications about that.
Please tell me of which cases you thought and I will do it.

asudoh replied:
Rather than talking about use-case, I had safe-guarding an exceptional situation in my mind. We can separate observation callback for `this.store` and each items in it and we don't need `Array.isArray()` check that may give us a false-positive.

harbalk replied:
Yes I thought about that maybe instead of testing if changeRecords[i].object is an array we could directly test if changeRecords[i].object === this.store ?
With that I will just have to add the condition of splice in the item callback of observe I think.

asudoh replied:
Well what I meant is:

```javascript
this.storeHandle = ObservableArray.observe(this.store, aCallback);

…

this.itemHandles[i] = Observable.observe(arrayItem, anotherCallback);
```

Given callbacks are not shared by above approach, you don’t need `Array.isArray()` check anymore.

harbalk replied:
Fixed in 17b07d7258e8832e8b56f2f0b39b864ed5f2f368.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+				if (this._observeResults) {
+					this._observeResults[0].deliver();
+					if (this._hasItemObservable) {
+						this.observeCallback && Observable.deliverChangeRecords(this.observeCallback);
```
Given [`Observable.observe()` call above](https://github.com/ibm-js/delite/pull/410/files#diff-3306122b521ab15fecaabf6fdc73a3dbR160) is done with what `Function.prototype.bind()` returns, this line delivers change records of a different function. You’ll need to keep around what `Function.prototype.bind()` returns and use it for `Observable.deliverChangeRecords()`.

harbalk replied:
I am not very familiar with the function bind.
Should I add bind to this.observeCallback in the call of deliverChangeRecords like that : 
```this.observeCallback && Observable.deliverChangeRecords(this.observeCallback.bind(this));```
Or add a bind to the deliverChangeRecords : 
```this.observeCallback && Observable.deliverChangeRecords(this.observeCallback).bind(this);```

asudoh replied:
What you need to know about `Function.prototype.bind()` in this context is that it returns a new (bound) function instead of changing the given function. In short:

```javascript
var fn = function () {};
alert(fn.bind({}) === fn); // false
```

harbalk replied:
Fixed in 17b07d7258e8832e8b56f2f0b39b864ed5f2f368.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+					this._beingDiscarded = true;
+					this._observeResults[0].deliver();
+					if (this._hasItemObservable === true) {
+						this.observeCallback && Observable.deliverChangeRecords(this.observeCallback);
```
Given [`Observable.observe()` call above](https://github.com/ibm-js/delite/pull/410/files#diff-3306122b521ab15fecaabf6fdc73a3dbR160) is done with what `Function.prototype.bind()` returns, this line delivers change records of a different function. You’ll need to keep around what `Function.prototype.bind()` returns and use it for `Observable.deliverChangeRecords()`.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+						// one item of the array modified
+					} else {
+						var found = false;
+						for (var j = i + 1; j < changeRecords.length; j++) {
```
Rather change doing two-dimensional array here,  you may do:

```javascript
var objects = [];
for (var i = 0; i < changeRecords.length; ++i) {
    var object = changeRecords[i].object;
    if (objects.indexOf(object) < 0) {
        objects.push(object);
        callback(object);
    }
}
```

Note that it just sends the object to the callback, given we only care about an object is changed or not.

The logic will be even simpler if we make sure only one object is handled in a callback, in such case it’d be like `callback(changeRecords[0].object)`.

harbalk replied:
Fixed in 17b07d7258e8832e8b56f2f0b39b864ed5f2f368. Thanks.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+		observeCallbackArray: function (change) {
+			if (change.type === "splice") {
+				var j;
+				for (j = 0; j < change.removed.length; j++) {
```
Observation of removed array items should be cleaned here.

harbalk replied:
Fixed in 17b07d7258e8832e8b56f2f0b39b864ed5f2f368

asudoh commented on Store.js:
```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+							evtAdded.index : this.renderItems.length;
+					}
+					// affect the callback to the observe function if the item is observable
+					Observable.observe(this.store[change.index + j], this.observeCallback.bind(this));
```
Should provide a way to stop observation (e.g. when the item is removed from array, when the widget is destroyed) of added array items.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,130 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+				} else {
+					this.itemHandles = [];
+					this.observeCallbackArray = this._observeCallbackArray.bind(this);
+					this.observeCallbackItems = this._observeCallbackItems.bind(this);
```
Did you mean to expose `this.observeCallbackArray`/`this.observeCallbackItems`?

asudoh commented on Store.js:
```diff
@@ -140,14 +151,130 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+					}
+					collection = processQueryResult.call(this, this.store.filter(this._isQueried, this));
+					// affect the callback to the observe function if the array is observable
+					this.storeHandle = ObservableArray.observe(this.store, this.observeCallbackArray);
```
Did you meant to expose `this.storeHandle`?

harbalk replied:
If you think it is better not to, I will modify the name with an underscore as for the observecallbacks. Maybe should I do it for the itemHandlers too ?

asudoh replied:
I thought both of them would be private, but it's a design choice whether you would want application developers to directly manipulate them.

asudoh commented on Store.js:
```diff
@@ -140,14 +151,130 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+					if (changeRecords[i].type === "splice") {
+						var j;
+						for (j = 0; j < changeRecords[i].removed.length; j++) {
+							this.itemHandles[changeRecords[i].index].remove();
```
Would you want to add `this.itemHandles.splice(changeRecords[i].index, 1)` so that indices of `this.itemHandles` will keep mapped to corresponding items in `this.store`?

asudoh commented on Store.js:
```diff
@@ -140,14 +151,130 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+									evtAdded.index : this.renderItems.length;
+							}
+							// affect the callback to the observe function if the item is observable
+							this.itemHandles[changeRecords[i].index + j] =
```
Would you want to splice in the new handle instead of (possibly) overwriting the handle of given index?
Also please make sure unit test covers handle cleanup stuffs, in case it hasn’t already.

harbalk replied:
Fixed in 238bd54542e54e0df450f0cd3b5ffb8c9d4f1bf3. Thanks a lot.

harbalk commented:
https://github.com/ibm-js/delite/pull/410#issuecomment-105873892
I tried to make the code more object oriented with this commit fd3fa31e946b3e3e5495f58e6cbdcb7c47aa1c87.

There is still two if {} else {} remaining. The one in computeProperties can not be avoid I think.
I can avoid the one in queryStoreAndInitItems by creating a queryStoreAndInitItems funtion in WrapObservableArray.js but I will lose some factorization.

Please let me know what you think about it.

harbalk commented:
Made two adapters in 679c119f8622351077f1e24b0d2a2473f35f28fb.

asudoh commented:
Thanks @harbalk. @cjolif @wkeese do we have a spec for “store adopter” BTW?

harbalk commented:
@asudoh here is a spec to describe the two adapters : https://docs.google.com/document/d/1M8uz6L9w3k85b0TYFTDTgF3oUZmVOsCLXybV0jBp3WI/edit?usp=sharing

asudoh commented:
Thanks @harbalk!

wkeese commented:
> Made two adapters in 679c119.

The thing is that your adapters have a `store` property that points back to the delite/Store object.   I meant for you to create classes that are self-contained.

harbalk commented:
@wkeese You want me to make the adapters heritate from the Store if I understand well. But it will completely modify the current use of Store.js, won't it ?
PS : @asudoh I have changed the permission.

wkeese commented:
Huh?  No, the adapters shouldn't extend the delite/Store class.  They probably shouldn't extend anything.

harbalk commented:
Ok then I don't see how I can acces to delite/Store functions like processCollection or initItems without referencing the store...

wkeese commented:
Exactly, the adapter shouldn't be calling delite/Store functions.  Rather, delite/Store will call functions in the adapter.

harbalk commented:
> Exactly, the adapter shouldn't be calling delite/Store functions. Rather, delite/Store will call functions in the adapter.

Make a new version in this commit : https://github.com/harbalk/delite/commit/d6879a37e2cd117fca482b6a7d0fee96f2e2f3b6

@wkeese what do you think about it ?

wkeese commented:
> Make a new version in this commit : harbalk@d6879a3

I don't understand, that version still seems to have the `store` property in ArrayToStoreAdapter.js.

harbalk commented:
> I don't understand, that version still seems to have the store property in ArrayToStoreAdapter.js.

Here the store property corresponds to the array or the dstore like you suggest in https://github.com/ibm-js/delite/pull/410#commitcomment-11645285. It is no more the delite/Store.

wkeese commented:
> Here the store property corresponds to the array or the dstore

I see.  You forgot to change line 33, which still says `@member {delite/Store}`.  But if `store` points to the array then I don't understand what `data` is for.  And also, I don't understand what `setCallback()` does?

`_storeHandle` is also a confusing name; perhaps `observeHandle` would be better.

Anyway, this is looking better than before.   I still don't understand it completely though.

harbalk commented:
> I see. You forgot to change line 33, which still says @member {delite/Store}.

Oh yes, sorry.

>  But if store points to the array then I don't understand what data is for.

The data corresponds to the store after passing in processQueryResult. After filtering with the query to be clear. (See https://github.com/ibm-js/delite/issues/308).
I need both because if there is a modification of an item in the store which is not in the data and if that modification make that item queried then I have to add it in the items displayed (renderItems) and also in the data.

> And also, I don't understand what setCallback() does?

setCallback is setting the functions to call when there is a modification on the array or on one of the item in the array.
Then for each item we associate the function _observeCallbackItems (for loop) and we keep the result to be able to remove the observation later.
And we associate the function _observeCallbackArray the same way.

> _storeHandle is also a confusing name; perhaps observeHandle would be better.

You are right, maybe _arrayHandle to match with _itemHandles ?

wkeese commented:
> setCallback is setting the functions to call when there is a modification on the array or on one of the item in the array.

OK, but that brings up a number of questions/issues:

1. `ArrayToStoreAdapter` extends `Evented`, and emits "add", "delete" etc. events.  It's redundant/inconsistent to also have a `setCallback()` method to do those same (or different?) notifications.
2. `DstoreToStoreAdapter` has an empty `setCallback()` method, so obviously it doesn't do anything.  That's inconsistent with  `ArrayToStoreAdapter`.
3. If you did decide to pass in callback methods, I think it would make more sense to pass them to the constructor.

harbalk commented:
> ArrayToStoreAdapter extends Evented, and emits "add", "delete" etc. events. It's redundant/inconsistent to also have a setCallback() method to do those same (or different?) notifications.

ArrayToStoreAdapter extends Evented because he has to notify to the delite/Store that there was a modification, because only the delite/Store can modify the itemRenderers with the function itemAdded, itemRemoved and itemUpdated. (I don't have access to this function in ArrayToStoreAdapter because I don't reference the delite/Store).

ObservableArray and specifically the observe function works with callback. There is no other way to use it.

The callbacks permit to make the communication between the Array and the Adapter and the events between the Adapter and the delite/Store.

> DstoreToStoreAdapter has an empty setCallback() method, so obviously it doesn't do anything. That's inconsistent with ArrayToStoreAdapter.

No it doesn't do anything because dstore/Store doesn't need callbacks. The dstore can directly communicate with the delite/Store (with events). There is no intermediate.

> If you did decide to pass in callback methods rather than emitting events, I think it would make more sense to pass them to the constructor.

I don't see why... The callbacks will always be the same and it is better to concentrate all the functions dedicated to ObservableArray in the ArrayToStoreAdapter, isn't it ?

asudoh commented:
@harbalk Any words on the comment I made to [here](https://docs.google.com/document/d/1M8uz6L9w3k85b0TYFTDTgF3oUZmVOsCLXybV0jBp3WI/edit) BTW?

wkeese commented:
I guess I misunderstood what you meant by:

> setCallback is setting the functions to call when there is a modification on the array or on one of the item in the array.

From looking at the code you are apparently passing the (observable) array to `setCallback()`.  First, it would be clearer if the method was called `setObservableArray()`, but the big question is, why do you need a `setCallback()` method when you are already passing the array to the constructor?

```js
new ArrayToStoreAdapter({store: this.store, query: this.query});
```

In other words, this combination looks redundant and therefore (apparently) poorly designed:

```js
this._storeAdapter = new ArrayToStoreAdapter({store: this.store, query: this.query});
...
this._storeAdapter.setCallback(this.store);
```

harbalk commented:
> @harbalk Any words on the comment I made to here BTW?
Oh yes sorry @asudoh I will complete the specifications. I am just waiting to validate the file with @wkeese.

For your second comment I don't know which is better. I made my implementation based on splice but it will be easy to modified if we need to.

harbalk commented:
@wkeese oh yes you're right at first I didn't save the array in the store field. I forgot to change the ```array``` parameter of the function to ```this.store```, I'm sorry.

wkeese commented:
OK.  It probably makes sense to complete the specification first so that it's easier for us to understand what these adapters do.

harbalk commented:
Made corrections on https://github.com/harbalk/delite/commit/a637871a5e39d7ce9203a5dbad7454cc9ddb1e6f.
I will correct the specification document soon.

wkeese commented:
> Made corrections on harbalk@a637871.

I don't think you need a `setObserveCallback()` public API method.  Just calling the ArrayToStoreAdapter()` constructor at the appropriate time should be enough.

harbalk commented:
> I don't think you need a setObserveCallback() public API method. Just calling the ArrayToStoreAdapter() constructor at the appropriate time should be enough.

Where do you think it is better to call the constructor ? In queryStoreAndInitItems ?

wkeese commented:
Sure, `queryStoreAndInitItems()` seems fine, assuming that you want to recreate the `ArrayToStoreAdapter()` every time the `query` changes (in addition to whenever the `store` changes).

harbalk commented:
Ok but it was also true where it was in computeProperties. Is there a big difference ?

wkeese commented:
`computeProperties()` seems fine too.   My point is to get rid of the `setObserveCallback()` method.

harbalk commented:
> My point is to get rid of the setObserveCallback() method.

Done in https://github.com/harbalk/delite/commit/da01cf93e90878eed29c76387f7312dd2f2cd1a9.

I have added comments of the functions in the files and I have redone the document of specifications.
https://docs.google.com/document/d/1M8uz6L9w3k85b0TYFTDTgF3oUZmVOsCLXybV0jBp3WI/edit?usp=sharing

wkeese commented:
> Done in harbalk@da01cf9.

OK thanks, I will check it out.

> I have added comments of the functions in the files and I have done the document of specifications again.

Thanks.  It looks like that document is explaining a lot of private internal methods?  All I really wanted to see is the API.  I.E. explanations of the methods that delite/Store will call.

harbalk commented:
Oh maybe I have to remove the underscores then. Because there are all functions that will be called by delite/Store of delite/StoreMap

wkeese commented:
It's hard to believe that delite/Store needs to call methods like _isQueried() and _redirectEvt().

harbalk commented:
I update the previous comment for _isQueried, but redirectEvt is called line 167 in delite/Store.

harbalk commented:
But maybe I can call it in the observe callbacks before sending the events. It was not possible before because I didn't have access to the collection but now that the collection is the adapter I can try.

wkeese commented:
OK.... well yes, I see that code in delite/Store now:

```js
					this._updateListener = collection.on("update", function (evt) {
						var idx = collection.data.indexOf(evt.obj);
						if (!this._storeAdapter._isAnUpdate(evt, idx)) {
							this._storeAdapter._redirectEvt(evt, idx);
						} else {
							evt = this._storeAdapter._updateItemInCollection(evt, idx);
							this._itemUpdated(evt);
						}
					}.bind(this));
```

I have no idea what the code is doing though, and I really hope it can be simplified.    For example, why does `collection` emit "update" events for things that are really adds or deletes?

harbalk commented:
All the modifications made on the items sent updates.

Like I said, before I have access to the collection only in the function queryStoreAndInitItems of delite/Store (before the adapters). So I could verify which kind of modifications it really was only there.

For example if I modify an item from the store which was not queried before but the modifications make him queried. Then it is not an update but finally an add if we consider the collection.

That's how an update could finally be an add or a remove.

But I think I can do it directly in the observe callbacks now. I'm working on that.

wkeese commented:
OK good.  The idea is to get the interface between classes as simple/clean as possible.

harbalk commented:
New commit : https://github.com/harbalk/delite/commit/8121b71134a4a4f63915eea781ad2b73b6791aec

I have moved the code which permit to update the collection and treat the event in ArrayToStoreAdapter.
I had to move the initialization of the adapters because the observe callback was called to early and make the tests failed.

I have completed a bit the _isQueried function to accept more filters.
@wkeese what do you think about it ?

wkeese commented:
harbalk@8121b71 looks a lot better.  I mainly looked at Store.js, but I can tell from that file that the interface to the adapters is much nicer than before.

About _isQueried(), I don't have much of an opinion about what kind of queries we should be supporting.  It's just something I noticed (that the array adapter was weaker than the dstore adapter) so I wanted to ask if that was intentional or not.

harbalk commented:
Ok, I tried to reproduce the same queries of dstore.

About the next step, the idea is to replace the use of deliteful/Store by the implementation done in this commit : https://github.com/harbalk/delite/commit/902b920cf747fcbbb91694aa94f2800b3dd064ef

Also it would be nice to rename the property store of the delite/Store at the same time. See https://github.com/ibm-js/delite/issues/308#issuecomment-115697421.

What do you think about it ?

asudoh commented:
Hi @harbalk sorry for being late on review, but I thought it'd be great if we can split store-specific part (`fetch()`, etc.) and observation part (`__observeCallbackItems()`, etc.). I'd see the primary purpose of the adopter a unified API for observing collection.

harbalk commented:
@asudoh I'm not sure what you are asking, did you mean to split the functions in different files or just make a better structure of the file to clearly see the functions dedicated to the observing and the ones store-specific ?

asudoh commented:
Hi @harbalk I think the former is better. Please feel free to let me know if you have any further questions. Thanks!

harbalk commented:
About which functions particularly were you talking ? On my side I think there is only five functions store-specific (fetch, fetchRange, setIdentity, get and getIdentity).

asudoh commented:
Hi @harbalk I think your list is mostly correct. To elaborate more on it, I think the adapter should focus on mutating and observing collection. Something like:

```javascript
var adapter = new CollectionAdapter(array /* or store */);
adapter.observe(function (records) { … }); // records is a splice type of change records
adapter.splice(1, 2, ‘a’, ‘b’); // Probably returning a promise considering async store
```

wkeese commented on Store.js:
```diff
@@ -102,6 +112,11 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+					this._storeAdapter = new DstoreToStoreAdapter({store: this});
+				} else {
+					this._storeAdapter = new ArrayToStoreAdapter({store: this});
+				}
```
Hmm, I just noticed that you are still passing in the the delite/Store as a property to the adapters.  As I've said before, the idea is for the adapters to be independent from the delite/Store code.   I thought you were passing in `this.store` not `this`.

harbalk replied:
I have done the modifications you asked in another branch. I haven't push it yet on the pull request that's why you can't see it here.

harbalk commented:
Update the PR with the new adapters and renaming of store property : https://github.com/harbalk/delite/commit/4652d7f2a0e00ce78ba5e68a62f832efe4111252

harbalk commented:
I have added a test on "new-query-asked" event and corrected the query in this commit : https://github.com/harbalk/delite/commit/00414f246aaf0f12012563418532c96d3d27c817.

Thanks @wkeese for your advices.

harbalk commented:
PR squashed in one commit to be pushed

wkeese commented:
I pushed the decor PR (moving ObservableArray from liaison to decor), and tried your latest squashed commit.  I'm getting one failure on both chrome/mac and FF/mac:

AllTypeOfQuery	(intermediate value).contains is not a function

So, please fix that.   Also, for documentation purposes the checkin comment should say "fixes #278".

harbalk commented:
Hi @wkeese, I don't reproduce the failure either on Saucelabs or in local. Can you tell me more about it ?

wkeese commented:
It's from the `new Filter().contains()` call.   Probably something w/the version of dstore.   Which version is it supposed to be?   Delite's bower.json says 1.0.x.

wkeese commented:
OK, it works with dstore 1.1, so I guess all the projects need to be updated to use 1.1.x.   Currently it's:

```
dapp/bower.json:		"dstore": "1.0.x"
delite/bower.json:		"dstore": "1.0.x",
deliteful/bower.json:		"dstore": "1.0.x"
dtreemap/bower.json:		"dstore": ">=0.0.1-dev"
```
I'll update delite and deliteful.

harbalk commented:
Hi @wkeese, I have made the modifications in this commit : https://github.com/harbalk/delite/commit/9f8c3bc098f3b6d59f2406a573785483422de9d4.
Not sure what I've done to remove the _getCollection function is fine with you, please tell me what you think about it.

wkeese commented:
> Not sure what I've done to remove the _getCollection function is fine with you, please tell me what you think about it.

That's the direction I was thinking of, but I just wasn't expecting to have a `this.collection` property.  For ArrayToStoreAdapter `this === this.collection`, so it doesn't make much sense.  Do you really need it for DstoreToStoreAdapter?

harbalk commented:
>  I think that ArrayToStoreAdapter and DstoreToStoreAdapter should represent the collection

Made a proposition in this commit : https://github.com/harbalk/delite/commit/199f035c7dde627fd64531e1c20b856325efa652

harbalk commented:
Updated with https://github.com/harbalk/delite/commit/70341c6628da3af7d15b9f8b3afafcc96b65bc7a.

wkeese commented:
> Updated with harbalk@70341c6.

OK thanks, this is looking pretty good.  Two remaining issues I noticed are:

1. jsdoc still not generated correctly.  All the members are listed as "<inner>".   (Try running `grunt jsdoc` and looking at `out/delite/docs/api/0.7.1/delite/ArrayToStoreAdapter.html`)
2. You have some methods with an underscore, indicating they are private, but they aren't.   They are used by the calling class.   For example, ArrayToStoreAdapter#_untrack() and ArrayToStoreAdapter#_setIdentity().   I guess you should just remove the underscore and not mark them as private.
3. When I run the regression now I get three errors on Chrome/mac in the Store-ObservableArray test.  How about you?   The errors are:

NullStore	Timeout reached on main - Store-ObservableArray - NullStore	0:01.507
Destroy	Cannot read property 'next_before' of undefined	0:00.003
SetNewStore	Timeout reached on main - Store-ObservableArray - SetNewStore	0:01.501

harbalk commented:
Hi @wkeese, I fixed the 3 points in : https://github.com/harbalk/delite/commit/9a1d06e07359db1fe7ea9c4c3373189b4723fda3.
Sorry for the third I thought I had run the tests.

wkeese commented on tests/unit/Store-ObservableArray.js:
```diff
@@ -0,0 +1,510 @@
+				//store.source.push(item);
+				//store.deliver();
+				//assert.deepEqual(store.renderItems[0], { id: "foo", name: "Foo", index: 0 });
+				//assert.deepEqual(store.renderItems[1], { id: "bar", name: "Bar", index: 1 });
```
why is all this commented out?

harbalk replied:
Because of the difference explain here : https://github.com/harbalk/delite/commit/1d9adf043fbf8547577d77bfb5a4dd5303318280#commitcomment-12636199.
Maybe I should remove the comments ?

wkeese replied:
Yes, either remove the commented out code, or add a comment that item adds/updates aren't expected to work in conjunction with a sort.

wkeese commented on tests/unit/Store-ObservableArray.js:
```diff
@@ -0,0 +1,510 @@
+		},
+
+		teardown: function () {
+			//container.parentNode.removeChild(container);
```
why is this commented out?   looks like it should be enabled.

harbalk replied:
It's not enabled in unit/Store.js. Should I enable it even so ?

harbalk replied:
No object container in the test, I should just remove the comment.

wkeese commented on tests/unit/Store-ObservableArray.js:
```diff
@@ -0,0 +1,510 @@
+			return d;
+		},
+
+		// TODO: re-enable when dstore will have re-introduced refresh event?
```
This comment seems strange, because this test file is not for dstore.

wkeese commented:
OK, thanks.  The tests seems to be passing now.

JSDoc is looking better but:

1. A number of methods in the *ToStoreAdapter are marked as protected.  Presumably they shouldn't be, since they are used by an external class rather than just by a subclass.
2. Should document the arguments passed to the constructor.   I guess that goes in the documentation for the class itself.
3. `data`, `source`, `query`, and `processQueryResult` should be marked as constant or readonly or private...  Actually, I'm not sure why you are listing those properties in the JSDoc at all.  The user never sets or accesses `data` directly, and the other properties only appear as arguments to the constructor (see (2)).
4. The types are not coming through.  For example, the JSDoc shows `data :Array` rather than `dstore/Collection`.  But that's not an issue anymore if you just get rid of `data` etc. from the JSDoc.

About the adapter code:

1. Is there any reason to have a `mix()` method?   Looks like it's only used internally, and only used once.
2. Do you really need the `dcl.after()` for the constructor? 

About the other code:

1. For StoreMap#preRender() etc., why is that code in StoreMap instead of Store?
2. The test for Store-ObservableArray.js is using dstore/Filter.   So even when using ObservableArray there's still a dependency on dstore?  That doesn't seem right.

harbalk commented:
> Do you really need the dcl.after() for the constructor? 

Issues appears when I remove it, I have reproduced the form of the other constructors I found in the different projects.

> The test for Store-ObservableArray.js is using dstore/Filter. So even when using ObservableArray there's still a dependency on dstore? That doesn't seem right.

See here https://github.com/ibm-js/delite/issues/278#issuecomment-131728433.

The other points are solved in : https://github.com/harbalk/delite/commit/01572880b6dcefa4c36eafc918923de6724e582e.

wkeese commented:
> The other points are solved in : harbalk@0157288.

OK good.  You should still get rid of the `dcl.after()` wrapper though, right?

wkeese commented:
Also, I'm confused by the `isQueried()` method.   It takes four arguments, but sometimes you are only calling it with one argument, for example:

```js
while (!this._isQueried(this.source[i])) {
```

What's going on there?

wkeese commented on docs/Store.md:
```diff
@@ -15,17 +15,19 @@ way of doing things.
 
-* store: references a `dstore/api/Store` implementation
+* source: references a `dstore/api/Store` implementation or an array
 * query: a query object to be passed to the store `filter()` function
```
Should say "or an array **or decor/ObservableArray**", right?

Also, the documentation on the `query` property needs to be updated because it's talking about the dstore `filter()` method, but that doesn't make sense if the user is specifying an array rather than a dstore.

asudoh commented on docs/Store.md:
```diff
@@ -53,16 +55,48 @@ render items using this mapping.
 
 By default those functions update the `renderItems` accordingly.
 
+#### Observable source with an array in source property
```
Good to have some sample code, showing e.g. how to update DOM upon change in collection.

harbalk replied:
Hi @asudoh, there is samples in deliteful PR if that's what you're asking.

wkeese replied:
There's also an example at the end of Store.md itself.  But I think @asudoh doesn't like it because it's using `refreshRendering()` rather than overriding `itemAdded()`, `itemUpdated()`, etc.

Note that the sample code is the same regardless of whether the source is an array or a dstore.

asudoh replied:
Thanks @harbalk @wkeese for clarification. As @wkeese implied what I'd like to see is samples in this .md (in addition to code in delite/samples). I'd like to see two things in the sample @wkeese  referred to:

* Make it more concrete, possibly with combo-box like widget or some other simple collection-based widget
* As @wkeese stated, examples with partial DOM update rather than updating the whole DOM for collection

cjolif replied:
If I'm not mistaken, what you are asking here is some enhancement of this documentation that would also have applied to the previous version (i.e. with dstore). In other words that is not specific to @harbalk work. Also knowing that @harbalk only have limited time until the end of his internship now (basically one week), I would suggest we skip this for now and we open a separate issue.

harbalk commented:
> Also, I'm confused by the isQueried() method. It takes four arguments, but sometimes you are only calling it with one argument, for example:

>while (!this._isQueried(this.source[i])) {

>What's going on there?

Hi @wkeese , this call is due to the filter function of Array. Normally it takes three arguments : the item, the index and the array. I only need the first one in my case.
But I had to pass a query to perform the `and` and `or` queries. Also I have to pass the query in the fourth arguments because the two previous are already set.

harbalk commented:
Fix doc and contructors in : https://github.com/harbalk/delite/commit/b559b11b4e71025dee1be42213ca4a7a1f2d75f2.

wkeese commented:
>  Normally it takes three arguments : the item, the index and the array. I only need the first one in my case.
> But I had to pass a query to perform the and and or queries. Also I have to pass the query in the fourth arguments because the two previous are already set.

Hmm, OK, that's very confusing to me.   This function is undocumented and sometimes it takes three arguments, sometimes it take four, and sometimes it takes one?   I don't understand the API to this function, and I'm wondering if it could be redesigned to be clearer.   At the least, it should be documented.

harbalk commented:
The function _isQueried is called with the filter function of Array here : 
```this.data = this.processQueryResult.call(this, this.source.filter(this._isQueried, this));```
As documented here https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter, the filter function automatically pass 3 arguments to his callback.

In every cases except the `and` and `or` queries, I don't need anything else than the item, that's why I put only the first argument when I call it out of the Array filter function.

In the case of an `and` and `or` queries, I have to give in arguments the first member of the `and` query (a query eq for example). But I can't pass it in the second argument because the filter function consider it as the index of the item. Then I have to pass it in the fourth place.

wkeese commented:
> The function _isQueried is called with the filter function of Array here

OK, I finally understand what's going on now.  But this seems like putting the cart before the horse.   If you changed:

```js
this.source.filter(this._isQueried, this)
```

to be:

```js
this.source.filter(function (item) { return this._isQueried(item); }.bind(this));
```

Then IIUC you can get rid of the bogus `index` and `tab` parameters to the `_isQueried()` and `_testItemProperty()` methods.

To be honest though, I think this code would work better (or at least faster) if the ArrayToStoreAdapter constructor compiled the query into a function.  I.E. make a recursive `_compileQuery()` method that takes a `query` Object and returns a corresponding function `f (item) --> true / false`.

harbalk commented:
> To be honest though, I think this code would work better (or at least faster) if the ArrayToStoreAdapter constructor compiled the query into a function. I.E. make a recursive _compileQuery() method that takes a query Object and returns a corresponding function f (item) --> true / false.

Done in https://github.com/harbalk/delite/commit/08e62aa792df32ef3bc71deaae60c71800ab7c1e.
Is that better ?

harbalk commented:
Updated with https://github.com/harbalk/delite/commit/bbdd4a13c48f5a4dd954d564570abecba02e309d.

wkeese commented on ArrayToStoreAdapter.js:
```diff
@@ -0,0 +1,382 @@
+
+		/**
+		 * Called to perform the fetchRange operation on the collection.
+		 * @param {args} - contains the start index and the end index of the fetch.
```
Looks like a typo, it should say something like `@param {Object} args -` (or something like that?

harbalk commented:
Updated with https://github.com/harbalk/delite/commit/b86862464843fa836a67f0b8cc40472bc2ec88dd.

wkeese commented:
OK, I pushed this code!  I made a number of corrections to JSDoc and some formatting fixes (see https://github.com/wkeese/delite/commit/ebe12f1b1fee5b451d668d0aa1041ae46f0a700a) but it was basically what's in this PR.

wkeese commented:
It would still be good to have a brief spec (I guess in the docs directory) for the DstoreToStoreAdapter / ArrayToStoreAdapter code.
